File: chatmessageview/src/main/java/com/github/bassaer/chatmessageview/models/Message.java
package com.github.bassaer.chatmessageview.models;

import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;

import com.github.bassaer.chatmessageview.utils.DateFormatter;
import com.github.bassaer.chatmessageview.utils.DefaultTimeFormatter;
import com.github.bassaer.chatmessageview.utils.IMessageStatusIconFormatter;
import com.github.bassaer.chatmessageview.utils.IMessageStatusTextFormatter;
import com.github.bassaer.chatmessageview.utils.ITimeFormatter;
import com.github.bassaer.chatmessageview.utils.SendTimeFormatter;

import java.util.Calendar;

/**
 * Message object
 * Created by nakayama on 2016/08/08.
 */
public class Message {

    /**
     * Sender information
     */
    private User mUser;

    /**
     * Whether sender username is shown or not
     */
    private boolean mUsernameVisibility = true;
    /**
     * If true, there is the icon space but invisible.
     */
    private boolean mIconVisibility = true;
    /**
     * If true, there is no icon space.
     */
    private boolean mHideIcon = false;

    /**
     * Whether the message is shown right side or not.
     */
    private boolean isRightMessage;

    /**
     * Message content text
     */
    private String mMessageText;

    /**
     * The time message that was created
     */
    private Calendar mCreatedAt;

    /**
     * Whether cell of list view is date separator text or not.
     */
    private boolean isDateCell;

    /**
     * TEXT format of the send time that is next to the message
     */
    private ITimeFormatter mSendTimeFormatter;

    /**
     * Date separator text format.
     * This text is shown if the before or after message was sent different day
     */
    private ITimeFormatter mDateFormatter;

    /**
     * Message status
     * You can use to know the message status such as fail, delivered, seen.. etc.
     */
    private int mStatus;

    /**
     * Message status is not shown.
     */
    public static final int MESSAGE_STATUS_NONE = 0;

    /**
     * Show message status icon.
     */
    public static final int MESSAGE_STATUS_ICON = 1;

    /**
     * Show message status text.
     * ex. seen, fail, delivered
     */
    public static final int MESSAGE_STATUS_TEXT = 2;

    /**
     * Show message status icon only right message
     */
    public static final int MESSAGE_STATUS_ICON_RIGHT_ONLY = 3;

    /**
     * Show message status icon only left message
     */
    public static final int MESSAGE_STATUS_ICON_LEFT_ONLY = 4;

    /**
     * Show message status text only right message
     */
    public static final int MESSAGE_STATUS_TEXT_RIGHT_ONLY = 5;

    /**
     * Show message status text only left message
     */
    public static final int MESSAGE_STATUS_TEXT_LEFT_ONLY = 6;

    /**
     * Message status type such as icon, text, or none.
     */
    private int mMessageStatusType = MESSAGE_STATUS_NONE;

    /**
     * Message status icon formatter
     */
    private IMessageStatusIconFormatter mStatusIconFormatter;

    /**
     * Message status text formatter
     */
    private IMessageStatusTextFormatter mStatusTextFormatter;

    /**
     * PICTURE message
     */
    private Bitmap mPicture;


    /**
     * Message Types
     *
     */
    public enum Type {
        TEXT,
        PICTURE,
        MAP,
        LINK
    }

    /**
     * Message type
     */
    private Type mType;

    /**
     * Constructor
     */
    public Message() {
        mCreatedAt = Calendar.getInstance();
        mSendTimeFormatter = new SendTimeFormatter();
        mDateFormatter = new DateFormatter();
        mSendTimeFormatter = new DefaultTimeFormatter();
        mType = Type.TEXT;
    }

    /**
     * Message builder
     */
    public static class Builder {
        private Message message;

        public Builder() {
            message = new Message();
        }

        public Builder setUser(User user) {
            message.setUser(user);
            return this;
        }

        public Builder setUsernameVisibility(boolean visibility) {
            message.setUsernameVisibility(visibility);
            return this;
        }

        public Builder setUserIconVisibility(boolean visibility) {
            message.setUsernameVisibility(visibility);
            return this;
        }

        public Builder hideIcon(boolean hide) {
            message.hideIcon(hide);
            return this;
        }


        public Builder setRightMessage(boolean isRight) {
            message.setRightMessage(isRight);
            return this;
        }

        public Builder setMessageText(String messageText) {
            message.setMessageText(messageText);
            return this;
        }

        public Builder setCreatedAt(Calendar calendar) {
            message.setCreatedAt(calendar);
            return this;
        }

        public Builder setDateCell(boolean isDateCell) {
            message.setDateCell(isDateCell);
            return this;
        }

        public Builder setSendTimeFormatter(ITimeFormatter sendTimeFormatter) {
            message.setSendTimeFormatter(sendTimeFormatter);
            return this;
        }

        public Builder setDateFormatter(ITimeFormatter dateFormatter) {
            message.setDateFormatter(dateFormatter);
            return this;
        }

        public Builder setStatus(int status) {
            message.setStatus(status);
            return this;
        }

        public Builder setMessageStatusType(int messageStatusType) {
            message.setMessageStatusType(messageStatusType);
            return this;
        }

        public Builder setStatusIconFormatter(IMessageStatusIconFormatter formatter) {
            message.setStatusIconFormatter(formatter);
            return this;
        }

        public Builder setStatusTextFormatter(IMessageStatusTextFormatter formatter) {
            message.setStatusTextFormatter(formatter);
            return this;
        }

        public Builder setType(Type type) {
            message.setType(type);
            return this;
        }

        public Builder setPicture(Bitmap picture) {
            message.setPicture(picture);
            return this;
        }

        public Message build() {
            return message;
        }

    }

    public User getUser() {
        return mUser;
    }

    public void setUser(User user) {
        mUser = user;
    }

    public boolean getUsernameVisibility() {
        return mUsernameVisibility;
    }

    public void setUsernameVisibility(boolean usernameVisibility) {
        mUsernameVisibility = usernameVisibility;
    }

    public boolean isIconHided() {
        return mHideIcon;
    }

    public void hideIcon(boolean hideIcon) {
        mHideIcon = hideIcon;
    }

    public boolean isRightMessage() {
        return isRightMessage;
    }

    public boolean getIconVisibility() {
        return mIconVisibility;
    }

    public void setIconVisibility(boolean iconVisibility) {
        mIconVisibility = iconVisibility;
    }

    public void setRightMessage(boolean isRightMessage) {
        this.isRightMessage = isRightMessage;
    }

    public String getMessageText() {
        return mMessageText;
    }

    public void setMessageText(String messageText) {
        mMessageText = messageText;
    }

    public Calendar getCreatedAt() {
        return mCreatedAt;
    }

    public void setCreatedAt(Calendar calendar) {
        mCreatedAt = calendar;
    }

    public String getTimeText() {
        return mSendTimeFormatter.getFormattedTimeText(mCreatedAt);
    }

    public boolean isDateCell() {
        return isDateCell;
    }

    public void setDateCell(boolean isDateCell) {
        this.isDateCell = isDateCell;
    }

    public String getDateSeparateText() {
        return mDateFormatter.getFormattedTimeText(mCreatedAt);
    }

    public int getStatus() {
        return mStatus;
    }

    public void setStatus(int status) {
        mStatus = status;
    }

    public int getMessageStatusType() {
        return mMessageStatusType;
    }

    public void setMessageStatusType(int messageStatusType) {
        mMessageStatusType = messageStatusType;
    }

    public IMessageStatusIconFormatter getStatusIconFormatter() {
        return mStatusIconFormatter;
    }

    public void setStatusIconFormatter(IMessageStatusIconFormatter statusIconFormatter) {
        mStatusIconFormatter = statusIconFormatter;
    }

    public Drawable getStatusIcon() {
        return mStatusIconFormatter.getStatusIcon(mStatus, isRightMessage());
    }

    public IMessageStatusTextFormatter getStatusTextFormatter() {
        return mStatusTextFormatter;
    }

    public void setStatusTextFormatter(IMessageStatusTextFormatter statusTextFormatter) {
        mStatusTextFormatter = statusTextFormatter;
    }

    public String getStatusText() {
        return mStatusTextFormatter.getStatusText(mStatus, isRightMessage());
    }

    public Type getType() {
        return mType;
    }

    public void setType(Type type) {
        mType = type;
    }

    public Bitmap getPicture() {
        return mPicture;
    }

    public void setPicture(Bitmap picture) {
        mPicture = picture;
    }

    /**
     * Set custom send time text formatter
     * @param sendTimeFormatter custom send time formatter
     */
    public void setSendTimeFormatter(ITimeFormatter sendTimeFormatter) {
        mSendTimeFormatter = sendTimeFormatter;
    }

    /**
     * Set custom date text formatter
     * @param dateFormatter custom date formatter
     */
    public void setDateFormatter(ITimeFormatter dateFormatter) {
        mDateFormatter = dateFormatter;
    }

    /**
     * Return Calendar to compare the day <br>
     * Reset hour, min, sec, milli sec.<br>
     * @return formatted calendar object
     */
    public Calendar getCompareCalendar() {
        Calendar calendar = (Calendar) mCreatedAt.clone();
        calendar.set(Calendar.HOUR, 0);
        calendar.set(Calendar.MINUTE, 0);
        calendar.set(Calendar.SECOND, 0);
        calendar.set(Calendar.MILLISECOND, 0);
        return calendar;
    }

    public interface OnBubbleClickListener {
        void onClick(Message message);
    }

    public interface OnBubbleLongClickListener {
        void onLongClick(Message message);
    }

    public interface OnIconClickListener {
        void onIconClick(Message message);
    }

    public interface OnIconLongClickListener {
        void onIconLongClick(Message message);
    }
}

File: chatmessageview/src/main/kotlin/com/github/bassaer/chatmessageview/model/Message.kt
package com.github.bassaer.chatmessageview.model

import android.graphics.Bitmap
import android.graphics.drawable.Drawable

import com.github.bassaer.chatmessageview.util.*

import java.util.*

/**
 * Message object
 * Created by nakayama on 2016/08/08.
 */
class Message {

    /**
     * Sender information
     */
    lateinit var user: IChatUser

    /**
     * Whether sender username is shown or not
     */
    var usernameVisibility = true
    /**
     * If true, there is the icon space but invisible.
     */
    var iconVisibility = true
    /**
     * If true, there is no icon space.
     */
    var isIconHided = false
        private set

    /**
     * Whether the message is shown right side or not.
     */
    var isRight: Boolean = false

    /**
     * Message content text
     */
    var text: String? = null

    /**
     * The time message that was created
     */
    var sendTime = Calendar.getInstance()

    /**
     * Whether cell of list view is date separator text or not.
     */
    var isDateCell: Boolean = false

    /**
     * TEXT format of the send time that is next to the message
     */
    private var mSendTimeFormatter: ITimeFormatter? = null

    /**
     * Date separator text format.
     * This text is shown if the before or after message was sent different day
     */
    private var mDateFormatter: ITimeFormatter? = null

    /**
     * Message status
     * You can use to know the message status such as fail, delivered, seen.. etc.
     */
    var status: Int = 0

    /**
     * Message status type such as icon, text, or none.
     */
    var statusStyle = STATUS_NONE

    /**
     * Message status icon formatter
     */
    var statusIconFormatter: IMessageStatusIconFormatter? = null

    /**
     *  Status Colorizer Enabled
     *  Can be used to disable the status icon colorizer so status icon keeps original color.
     *  Use when updating status icon to another drawable and you don't want it's color to be altered
     */
    var statusColorizeEnabled: Boolean = true

    /**
     * Message status text formatter
     */
    var statusTextFormatter: IMessageStatusTextFormatter? = null

    /**
     * PICTURE message
     */
    var picture: Bitmap? = null

    /**
     * Message type
     */
    var type: Type? = null

    val timeText: String
        get() = mSendTimeFormatter!!.getFormattedTimeText(sendTime!!)

    val dateSeparateText: String
        get() = mDateFormatter!!.getFormattedTimeText(sendTime!!)

    val statusIcon: Drawable
        get() = statusIconFormatter!!.getStatusIcon(status, isRight)

    val statusText: String
        get() = statusTextFormatter!!.getStatusText(status, isRight)

    /**
     * Message Types
     *
     */
    enum class Type {
        TEXT,
        PICTURE,
        MAP,
        LINK
    }

    /**
     * Constructor
     */
    init {
        sendTime = Calendar.getInstance()
        mSendTimeFormatter = DefaultTimeFormatter()
        mDateFormatter = DateFormatter()
        mSendTimeFormatter = DefaultTimeFormatter()
        type = Type.TEXT
    }

    /**
     * Message builder
     */
    class Builder {
        private val message: Message = Message()

        fun setUser(user: IChatUser): Builder {
            message.user = user
            return this
        }

        fun setUsernameVisibility(visibility: Boolean): Builder {
            message.usernameVisibility = visibility
            return this
        }

        fun setUserIconVisibility(visibility: Boolean): Builder {
            message.iconVisibility = visibility
            return this
        }

        fun hideIcon(hide: Boolean): Builder {
            message.hideIcon(hide)
            return this
        }


        fun setRight(isRight: Boolean): Builder {
            message.isRight = isRight
            return this
        }

        fun setText(text: String): Builder {
            message.text = text
            return this
        }

        fun setSendTime(calendar: Calendar): Builder {
            message.sendTime = calendar
            return this
        }

        fun setDateCell(isDateCell: Boolean): Builder {
            message.isDateCell = isDateCell
            return this
        }

        fun setSendTimeFormatter(sendTimeFormatter: ITimeFormatter): Builder {
            message.setSendTimeFormatter(sendTimeFormatter)
            return this
        }

        fun setDateFormatter(dateFormatter: ITimeFormatter): Builder {
            message.setDateFormatter(dateFormatter)
            return this
        }

        fun setStatus(status: Int): Builder {
            message.status = status
            return this
        }

        fun setStatusStyle(statusStyle: Int): Builder {
            message.statusStyle = statusStyle
            return this
        }

        fun setStatusIconFormatter(formatter: IMessageStatusIconFormatter): Builder {
            message.statusIconFormatter = formatter
            return this
        }

        fun setStatusTextFormatter(formatter: IMessageStatusTextFormatter): Builder {
            message.statusTextFormatter = formatter
            return this
        }

        fun setType(type: Type): Builder {
            message.type = type
            return this
        }

        fun setPicture(picture: Bitmap): Builder {
            message.picture = picture
            return this
        }

        fun build(): Message {
            return message
        }

    }

    fun hideIcon(hideIcon: Boolean) {
        isIconHided = hideIcon
    }

    /**
     * Set custom send time text formatter
     * @param sendTimeFormatter custom send time formatter
     */
    fun setSendTimeFormatter(sendTimeFormatter: ITimeFormatter) {
        mSendTimeFormatter = sendTimeFormatter
    }

    /**
     * Set custom date text formatter
     * @param dateFormatter custom date formatter
     */
    fun setDateFormatter(dateFormatter: ITimeFormatter) {
        mDateFormatter = dateFormatter
    }

    interface OnBubbleClickListener {
        fun onClick(message: Message)
    }

    interface OnBubbleLongClickListener {
        fun onLongClick(message: Message)
    }

    interface OnIconClickListener {
        fun onIconClick(message: Message)
    }

    interface OnIconLongClickListener {
        fun onIconLongClick(message: Message)
    }

    companion object {

        /**
         * Message status is not shown.
         */
        val STATUS_NONE = 0

        /**
         * Show message status icon.
         */
        val STATUS_ICON = 1

        /**
         * Show message status text.
         * ex. seen, fail, delivered
         */
        val STATUS_TEXT = 2

        /**
         * Show message status icon only right message
         */
        val STATUS_ICON_RIGHT_ONLY = 3

        /**
         * Show message status icon only left message
         */
        val STATUS_ICON_LEFT_ONLY = 4

        /**
         * Show message status text only right message
         */
        val STATUS_TEXT_RIGHT_ONLY = 5

        /**
         * Show message status text only left message
         */
        val STATUS_TEXT_LEFT_ONLY = 6
    }
}
